### PR TITLE
DBZ-2027 Upgrading Postgres JDBC driver to 42.2.12

### PR DESCRIPTION
**Pull request by gunnarmorling**
https://issues.redhat.com/browse/DBZ-2027

### CODE REVIEW

**Comment by gunnarmorling** [comment]
```
java.util.NoSuchElementException: No value present
	at java.util.Optional.get(Optional.java:135)
	at io.debezium.connector.postgresql.PostgresValueConverter.numericSchema(PostgresValueConverter.java:301)
	at io.debezium.connector.postgresql.PostgresValueConverter.schemaBuilder(PostgresValueConverter.java:192)
```

That one is interesting, it indicates a behavioral change around NUMERIC columns without length and scale. They used to no scale and length set to -1, whereas now with that driver update length is set to 131089. // CC @davecramer, was that an intentional change? Just curiosity on my side, I think we can handle it.

**Comment by gunnarmorling** [comment]
@jpechane, can you check out this one? Let's see how the next CI run goes, but I think remaining failures are unrelated timing issues in the test suite.

**Comment by gunnarmorling** [comment]
The change related to scale was introduced by https://github.com/pgjdbc/pgjdbc/issues/1767; see the first commit message in this PR for details.

**Comment by gunnarmorling** [comment]
And another behavioural change in the driver (see https://github.com/pgjdbc/pgjdbc/pull/1708).

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/RecordsSnapshotProducerIT.java [hunk 1]
```diff
@@ -434,7 +435,7 @@ public void shouldGenerateSnapshotsForPartitionedTables() throws Exception {
 
         // verify each topic contains exactly the number of input records
         assertEquals(1, topicCounts.get("test_server.public.first_table").intValue());
-        assertEquals(30, topicCounts.get("test_server.public.partitioned").intValue());
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/RecordsSnapshotProducerIT.java:437]: @jpechane, WDYT about this one? I think it's ok to include this change in a minor; I'd argue it makes more sense that way actually. But it also settles that we shouldn't add this in 1.1.x IMO.

**Comment by gunnarmorling** [comment]
Sigh, something still is broken. @jpechane, should you have any idea regarding the "Unknown type named timestamp without time zone requested" issue in the logs, I'll be all ears :)

**Comment by Naros** [comment]
@gunnarmorling I'm wondering if the driver change is now giving us some variant column types that were didn't get previously and perhaps `TypeRegistry#get(String)` needs to account for this now.  What if you changed the specific line:
```
PostgresType r = nameToType.get(name);
```
to 
```
PostgresType r = nameToType.get(normalizeTypeName(name));
```
It looks like that would be backward compatible I believe.

**Comment by davecramer** [comment]
are you using 42.2.11 or 42.2.12? This makes sense for .11, but not .12

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/RecordsSnapshotProducerIT.java (hunk 1 above)
Comment (jpechane) [reply debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/RecordsSnapshotProducerIT.java:437]: @gunnarmorling That's definiteyl breaking compatibility change so no way adding it to 1.1.x. IIUC this now aligns Postres with MySQL and we must warn the user to use `ByLogicalTableRouter` to get the functionality back.

**Comment by jpechane** [comment]
@davecramer  We are using `11` in this PR. Should we use `12` and the issue will be fixed?

**Comment by jpechane** [comment]
@gunnarmorling I am quite unhappy with the current state of affairs. IMHO with things as they are now we should introduce compatibility matrix in the docs not only for database versions but also for JDBC driver versions.

**Comment by gunnarmorling** [comment]
> Should we use 12 and the issue will be fixed?

Yes, we should use 42.2.12; 42.2.11 is borked as per Dave. And in fact I was using 42.2.12, but then went back (temporarily, or so I hoped) to the earlier one to understand when the partition change came in. Just rebased and force-pushed to use 42.2.12 again.

**Comment by gunnarmorling** [comment]
> we should introduce compatibility matrix in the docs not only for database versions but also for JDBC driver versions.

@jpechane, yes, sounds like a good idea to add the driver versions. Can you log an issue?

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/RecordsSnapshotProducerIT.java (hunk 1 above)
Comment (gunnarmorling) [reply debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/RecordsSnapshotProducerIT.java:437]: Hum, hum. So is MySQL indeed creating one topic per partition? Wondering whether it'd not make more sense to write all partitions to a single topic actually; is the information on partitioning relevant for downstream consumers?

Comment (jpechane) [reply debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/RecordsSnapshotProducerIT.java:437]: @gunnarmorling I stand corrected. MySQL partitioning is about spreading the table into different physical files. POstgresSQL partitioning is about spreading to different tables. MySQL equivalent is sharding.

Comment (gunnarmorling) [reply debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/RecordsSnapshotProducerIT.java:437]: Ok, so what number of topics do we get for a partitioned table in MySQL? Is it the same as for Postgres with the driver update?

Comment (jpechane) [reply debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/RecordsSnapshotProducerIT.java:437]: @gunnarmorling If I understand this change then for n paritions and one aggregate table we get
1) n tables for PostgreSQL partitioned table
2) 1 table for MySQL partitioned table
3) n table for MySQL sharded tables

As PostgreSQL partitioning seems to be equivalent to MySQL sharding

**Comment by jpechane** [comment]
@gunnarmorling Applied, thanks for the thorough analysis.

**Pull request merged**